src/logic/hosts.rs

```diff
@@ -0,0 +1,69 @@
+use std::collections::HashSet;
+
+use chrono::Utc;
+use thiserror::Error;
+
+use crate::db::{
+    models::{Host, HostId, LeasedHost, UserId},
+    Registry,
+};
+
+#[derive(Error, Debug)]
+pub enum HostError {
+    #[error("Database error")]
+    DatabaseError(#[from] sqlx::Error),
+
+    #[error("Host is already leased")]
+    AlreadyLeased(Vec<HostId>),
+}
+
+pub async fn get_available_hosts(registry: &Registry) -> Result<Vec<Host>, HostError> {
```

Comment:
"HostError" contain "AlreadyLeased" error but this function will never got this error

Reply:
It can get `DatabaseError` variant though. The idea is to have the domain-level error type, to make it easier to implement error handling on the web server level. 

It'll be revised many times, in particular I don't like the name